File: Shajee_CS280 HW#2/main.cpp
#include <fstream>
#include <iostream>
#include <string>

using std::cout;
using std::endl;
using std::ifstream;
using std::ios;
using std::ofstream;
using std::string;
string line;
int i;
bool A(void);
bool I(void);
bool E(void);
bool P(void);
bool O(void);
bool L(void);
bool U(void);
bool C(void);
bool D(void);
bool A(){
      if(I()){
         if(line[i]=='='){
            i++;
            if(E()){
               return true;
            }
         }
         return false;
      }
      return false;
   }
bool E(){
      if(P()){
         if(O()){
            if(P()){
               return true;
            }
            return false;
         }
         return true;
      }
      return false;
   }
bool O(){
      if(i < line.length()&&(line[i]=='+'||line[i]=='-'||line[i]=='/')){
         i++;
         return true;
      }
      else if(i < line.length()&&line[i]=='*'){
         i++;
         if(i < line.length()&&line[i]=='*'){
            i++;
            return true;
         }
         return true;
      }
      return false;
   }
bool P(){
      if(I()){
         return true;
      }
      else{
         if(L()){
            return true;
         }
         else{
            if(U()){
               if(I()){
                  return true;
               }
               else{
                  if(L()){
                     return true;
                  }
                  else{
                     return false;
                  }
               }
            }
            else{
               if(i < line.length()&&line[i]=='('){
                  i++;
                  if(E()){
                     if(i < line.length()&&line[i]==')'){
                        i++;
                        return true;
                     }
                  }
                  return false;
               }
               return false;
            }
         }
      }
   }
bool U()
        {
        	if (i<line.length()&&(line[i] == '+' || line[i] == '-' || line[i] == '!')) 
        	{
            ++i;
        		return true;
        		}
        	return false;
        	}
bool I(){
      if(C()){
         if(I()&&i<line.length()){
            return true;
         }
         return true;
      }
      return false;
   }
bool C(){
      if(i<line.length()&&(line[i]>='a'&&line[i]<='z')){
         ++i;
         return true;
      }
      return false;
   }
bool L(){
      if(D()){
         if(L()&&i<line.length()){
            return true;
         }
         return true;
      }
      return false;
   }
bool D(){
      if(i<line.length()&&(line[i]>='0'&&line[i]<='9')){
         ++i;
         return true;
      }
      return false;
   }
int main(void) {

    ifstream fin("input.txt");
    ofstream fout("output.txt", ios::out|ios::app);


    while (fin >> line) {
       i=0;
       if(A()&&line[i] == '\0'){
         fout<<"The string \""<<line<<"\" is in the language."<<endl;
       }
       else{
         fout<<"The string \""<<line<<"\" is not in the language."<<endl;
       }
    }
    fout.close();
    fin.close();

    return 0;
}

File: Shajee_CS280 HW#2/Main.java
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

public class Main {
   public static String line;
   public static int i;
   public static boolean A(){
      if(I()){
         if(line.charAt(i)=='='){
            i++;
            if(E()){
               return true;
            }
         }
         return false;
      }
      return false;
   }
   public static boolean E(){
      if(P()){
         if(O()){
            if(P()){
               return true;
            }
            return false;
         }
         return true;
      }
      return false;
   }
   public static boolean O(){
      if(i < line.length()&&(line.charAt(i)=='+'||line.charAt(i)=='-'||line.charAt(i)=='/')){
         i++;
         return true;
      }
      else if(i < line.length()&&line.charAt(i)=='*'){
         i++;
         if(i < line.length()&&line.charAt(i)=='*'){
            i++;
            return true;
         }
         return true;
      }
      return false;
   }
   public static boolean P(){
      if(I()){
         return true;
      }
      else{
         if(L()){
            return true;
         }
         else{
            if(U()){
               if(I()){
                  return true;
               }
               else{
                  if(L()){
                     return true;
                  }
                  else{
                     return false;
                  }
               }
            }
            else{
               if(i < line.length()&&line.charAt(i)=='('){
                  i++;
                  if(E()){
                     if(i < line.length()&&line.charAt(i)==')'){
                        i++;
                        return true;
                     }
                  }
                  return false;
               }
               return false;
            }
         }
      }
   }
   public static boolean U()
        {
        	if (i<line.length()&&(line.charAt(i) == '+' || line.charAt(i) == '-' || line.charAt(i) == '!')) 
        	{
            ++i;
        		return true;
        		}
        	return false;
        	}
   public static boolean I(){
      if(C()){
         if(I()){
            return true;
         }
         return true;
      }
      return false;
   }
   public static boolean C(){
      if(i<line.length()&&(line.charAt(i)>='a'&&line.charAt(i)<='z')){
         ++i;
         return true;
      }
      return false;
   }
   public static boolean L(){
      if(D()){
         if(L()){
            return true;
         }
         return true;
      }
      return false;
   }
   public static boolean D(){
      if(i<line.length()&&(line.charAt(i)>='0'&&line.charAt(i)<='9')){
         ++i;
         return true;
      }
      return false;
   }
   public static void main(String [] args){
        File in = new File("input.txt");
        File out = new File("output.txt");

        try (Scanner scan = new Scanner(in);
             PrintWriter pw = new PrintWriter(new FileWriter(out, true))) {

            while (scan.hasNextLine()) {
                line=scan.nextLine();
                i=0;
                if(A() && i==line.length()){
                  pw.println("The string \""+line+"\" is in the language.");
                }
                else{
                  pw.println("The string \""+line+"\" is not in the language.");
                }
            }
        }
        catch (IOException e) {
            e.printStackTrace();
            System.exit(0);
        }

   }
}